src/expr/symbol_table.cpp

```diff
@@ -41,23 +42,165 @@ using ::std::pair;
+  */
+  Expr getOverloadedConstantForType(const std::string& name, Type t) const;
+  
+  /**
+   * If possible, returns a defined function for a name
+   * and a vector of expected argument types. Otherwise returns
+   * null expression.
+   */
+  Expr getOverloadedFunctionForTypes(const std::string& name, const std::vector< Type >& argTypes) const;
+  /** called when obj is bound to name, and prev_bound_obj was already bound to name */
+  void bind(const string& name, Expr prev_bound_obj, Expr obj);
+private:
+  /** Marks expression obj with name as overloaded. 
+   * Adds relevant information to the type arg trie data structure.
+  */
+  void markOverloaded(const string& name, Expr obj);
+  /** the null expression */
+  Expr d_nullExpr;
+  // The (context-independent) trie storing that maps expected argument
+  // vectors to symbols. All expressions stored in d_symbols are only 
+  // interpreted as active if they also appear in the context-dependent
+  // set d_overloaded_symbols.
+  class TypeArgTrie {
+  public:
+    // children of this node
+    std::map< Type, TypeArgTrie > d_children;
+    // symbols at this node
+    std::map< Type, Expr > d_symbols;
+  };
+  /** for each string with operator overloading, this stores the data structure above. */
+  std::unordered_map< std::string, TypeArgTrie > d_overload_type_arg_trie;
+  /** The set of overloaded symbols. */
+  CDHashSet<Expr, ExprHashFunction>* d_overloaded_symbols;
+};
+
+bool OverloadedTypeTrie::isOverloadedFunction(Expr fun) const {
+  return d_overloaded_symbols->find(fun)!=d_overloaded_symbols->end();
+}
+
+Expr OverloadedTypeTrie::getOverloadedConstantForType(const std::string& name, Type t) const {
+  std::unordered_map< std::string, TypeArgTrie >::const_iterator it = d_overload_type_arg_trie.find(name);
+  if(it!=d_overload_type_arg_trie.end()) {
+    std::map< Type, Expr >::const_iterator its = it->second.d_symbols.find(t);
+    if(its!=it->second.d_symbols.end()) {
+      return its->second;
+    }
+  }
+  return d_nullExpr;
+}
+
+Expr OverloadedTypeTrie::getOverloadedFunctionForTypes(const std::string& name, 
+                                                       const std::vector< Type >& argTypes) const {
+  std::unordered_map< std::string, TypeArgTrie >::const_iterator it = d_overload_type_arg_trie.find(name);
+  if(it!=d_overload_type_arg_trie.end()) {
+    const TypeArgTrie * tat = &it->second;
+    for(unsigned i=0; i<argTypes.size(); i++) {
+      std::map< Type, TypeArgTrie >::const_iterator itc = tat->d_children.find(argTypes[i]);
+      if(itc!=tat->d_children.end()) {
+        tat = &itc->second;
+      }else{
+        // no functions match 
+        return d_nullExpr;
+      }
+    }
+    // now, we must ensure that there is one and only one active symbol at this node
+    Expr retExpr;
+    for(std::map< Type, Expr >::const_iterator its = tat->d_symbols.begin(); its != tat->d_symbols.end(); ++its) {
+      Expr expr = its->second;
```

Comment:
Did you try this:
for (Expr expr : tat->d_symbols)

These range-based for loops make the code a lot shorter and IMO easier to read.  I'd like to encourage using them.

Reply:
We can discuss this in the coding guidelines. In my opinion, I think it is helpful to see the type of the iterator and makes the code more robust.  Say I later update the type of d_symbols, I want the code to tell me all places where I need to reconsider what I am doing with d_symbols.